src/Scene/Scene.cpp

```diff
@@ -51,33 +51,73 @@ GameObject* Scene::NewObject(std::string name)
-                               // Setting reCalculateLocalTQS to false will prevent unnecessary calculations
-                               pGameObject->GetTransform()->SetParent(nullptr, false);
-                           }
-                           if (pGameObject == gameObject || pGameObject->GetTransform()->IsChildOf(gameObject->GetTransform(), true))
-                           {
-                               LOG_STREAM(DEBUG) << "Destroying GameObject " << pGameObject->GetName() << std::endl;
-                               delete pGameObject;
-                               return true;
-                           }
-                           return false; });
-
-    LOG_STREAM(DEBUG) << "Destroyed " << (std::distance(newLogicalEnd, m_GameObjects.end()))
-                      << " GameObjects" << std::endl;
-    m_GameObjects.erase(newLogicalEnd, m_GameObjects.end());
+    DestroyMultipleObjects(&pGameObject, 1);
+}
+
+void Scene::DestroyMultipleObjects(GameObject** ppGameObjects, int count)
+{
+    if (ppGameObjects == nullptr)
+    {
+        LOG_STREAM(ERROR) << "Cannot destroy nullptr" << std::endl;
+        return;
+    }
+
+    // dfs by a while loop
+    std::vector<GameObject*> stack;
+    for (int i = 0; i < count; ++i)
+    {
+        if (ppGameObjects[i] == nullptr)
+        {
+            LOG_STREAM(ERROR) << "Cannot destroy nullptr" << std::endl;
+            continue;
+        }
+        if (ppGameObjects[i]->GetScene() != this)
+        {
+            LOG_STREAM(ERROR) << "Cannot destroy object" << ppGameObjects[i]->GetName() << "from another scene." << std::endl;
+            continue;
+        }
+        stack.push_back(ppGameObjects[i]);
+        ppGameObjects[i]->GetTransform()->SetParent(nullptr, false);
+    }
+
+    while (!stack.empty())
+    {
+        auto* current = stack.back();
+        stack.pop_back();
+
+        for (auto const& child : *(current->GetTransform()))
+        {
+            // m_GameObjectsToDestroy is a comes-for-free visited set, by which the maximum iteration count will be the number of game objects
+            if (!m_GameObjectsToDestroy[child->GetGameObject()])
+            {
+                stack.push_back(child->GetGameObject());
+            }
+        }
+
+        m_GameObjectsToDestroy[current] = true;
+    }
+
+    auto destroyedGameObjectCount = std::erase_if(m_GameObjects, [this](auto& go)
+                                                  { return m_GameObjectsToDestroy[go]; });
+
+    // Now it's time to delete the game object hard
+    std::for_each(m_GameObjectsToDestroy.begin(), m_GameObjectsToDestroy.end(), [](auto& pair)
+                  {if (pair.second)
+                      {
+                          LOG_STREAM(DEBUG) << "Destroying " << pair.first->GetName() << std::endl;
+                          delete pair.first;
+                      } });
+
+    std::erase_if(m_GameObjectsToDestroy, [](auto& pair)
```

Comment:
Why make `m_GameObjectsToDestroy` a member variable instead of a temporary variable? It is used only inside Destroy and has no other use, and there's no need to cache anything inside of it. It should be OK to just leave it here if it's a temp var.

Also how about using `unordered_set` instead of `unordered_map`? We can remove the usage of the boolean value just by using `contains` which has the same complexity as `operator[]`.

Reply:
> Why make m_GameObjectsToDestroy a member variable instead of a temporary variable?

To be honest, I did spend quite some time hesitating if we should use a local temp variable or a data field, and still do not have a strong reason on each side.
Making it a data field looks like a very first (and humble) prototype version, or something to start with, if by any chances we want to make it support things like "change list / command buffer". Like, If something were to be destroyed, we give it a mark/tag, then (presumably at the same frame, before everything submitted to the rendering system) another guy will have an easy way to know if a GameObject they want access has been destroyed.


> Also how about using unordered_set instead of unordered_map?
You are right. `unordered_set` is the better choice, especially if we are to use only a temp var

Reply:
I believe the better way to implement a deferred deletion is to add a bool field to the GameObject itself marking its status instead of using a container in Scene that requires maintenance.

Reply:
Switched to `unordered_set` and used temp var instead. Changes submitted